src/lib/__tests__/InstantSearch-integration-test.ts

```diff
@@ -44,3 +46,72 @@ describe('configure', () => {
     });
   });
 });
+
+describe('middleware', () => {
+  it("runs middlewares' onStateChange uiState has changed", async () => {
+    const container = document.createElement('div');
+    const search = instantsearch({
+      indexName: 'instant_search',
+      searchClient: createSearchClient(),
+    });
+
+    const middlewareDefinition: MiddlewareDefinition = {
+      onStateChange: jest.fn(),
+      subscribe: jest.fn(),
+      unsubscribe: jest.fn(),
+    };
+
+    search.use(() => middlewareDefinition);
+
+    search.addWidgets([
+      searchBox({
+        container,
+        placeholder: 'search',
+      }),
+    ]);
+
+    search.start();
+
+    fireEvent.input(container.querySelector('input')!, {
+      target: { value: 'q' },
+    });
+
+    await runAllMicroTasks();
+    expect(middlewareDefinition.onStateChange).toHaveBeenCalledTimes(1);
+  });
+
+  it("runs middlewares' onStateChange uiState has changed and onStateChange was provided", async () => {
```

Comment:
I don't understand this test name. Can we rephrase it?

Reply:
```suggestion
  it("runs middlewares' onStateChange when uiState changes (with root onStateChange)", async () => {
```

Hard to phrase it correctly, but this is the bug fixed in this PR. Is the suggestion better @francoischalifour ?

Reply:
Better yes. How about this?

```suggestion
  it("runs middlewares' onStateChange when uiState changes with user-provided onStateChange param", async () => {
```